File: JsonSchemaAndStructs/KiteConnect-Python-DataClass/quote.py
# This code parses date/times, so please
#
#     pip install python-dateutil
#
# To use this code, make sure you
#
#     import json
#
# and then, to convert JSON from a string, do
#
#     result = quote_from_dict(json.loads(json_string))

from dataclasses import dataclass
from typing import Optional, Any, List, Dict, TypeVar, Callable, Type, cast
from datetime import datetime
import dateutil.parser


T = TypeVar("T")


def from_int(x: Any) -> int:
    assert isinstance(x, int) and not isinstance(x, bool)
    return x


def from_none(x: Any) -> Any:
    assert x is None
    return x


def from_union(fs, x):
    for f in fs:
        try:
            return f(x)
        except:
            pass
    assert False


def from_float(x: Any) -> float:
    assert isinstance(x, (float, int)) and not isinstance(x, bool)
    return float(x)


def to_float(x: Any) -> float:
    assert isinstance(x, float)
    return x


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


def from_datetime(x: Any) -> datetime:
    return dateutil.parser.parse(x)


def from_dict(f: Callable[[Any], T], x: Any) -> Dict[str, T]:
    assert isinstance(x, dict)
    return { k: f(v) for (k, v) in x.items() }


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


@dataclass(slots=True)
class Buy:
    orders: Optional[int] = None
    price: Optional[float] = None
    quantity: Optional[int] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Buy':
        assert isinstance(obj, dict)
        orders = from_union([from_int, from_none], obj.get("orders"))
        price = from_union([from_float, from_none], obj.get("price"))
        quantity = from_union([from_int, from_none], obj.get("quantity"))
        return Buy(orders, price, quantity)

    def to_dict(self) -> dict:
        result: dict = {}
        result["orders"] = from_union([from_int, from_none], self.orders)
        result["price"] = from_union([to_float, from_none], self.price)
        result["quantity"] = from_union([from_int, from_none], self.quantity)
        return result


@dataclass(slots=True)
class Depth:
    buy: Optional[List[Buy]] = None
    sell: Optional[List[Buy]] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Depth':
        assert isinstance(obj, dict)
        buy = from_union([lambda x: from_list(Buy.from_dict, x), from_none], obj.get("buy"))
        sell = from_union([lambda x: from_list(Buy.from_dict, x), from_none], obj.get("sell"))
        return Depth(buy, sell)

    def to_dict(self) -> dict:
        result: dict = {}
        result["buy"] = from_union([lambda x: from_list(lambda x: to_class(Buy, x), x), from_none], self.buy)
        result["sell"] = from_union([lambda x: from_list(lambda x: to_class(Buy, x), x), from_none], self.sell)
        return result


@dataclass(slots=True)
class Ohlc:
    close: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    open: Optional[int] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Ohlc':
        assert isinstance(obj, dict)
        close = from_union([from_float, from_none], obj.get("close"))
        high = from_union([from_float, from_none], obj.get("high"))
        low = from_union([from_float, from_none], obj.get("low"))
        open = from_union([from_int, from_none], obj.get("open"))
        return Ohlc(close, high, low, open)

    def to_dict(self) -> dict:
        result: dict = {}
        result["close"] = from_union([to_float, from_none], self.close)
        result["high"] = from_union([to_float, from_none], self.high)
        result["low"] = from_union([to_float, from_none], self.low)
        result["open"] = from_union([from_int, from_none], self.open)
        return result


@dataclass(slots=True)
class Datum:
    average_price: Optional[float] = None
    buy_quantity: Optional[int] = None
    depth: Optional[Depth] = None
    instrument_token: Optional[int] = None
    last_price: Optional[float] = None
    last_quantity: Optional[int] = None
    last_trade_time: Optional[datetime] = None
    lower_circuit_limit: Optional[float] = None
    net_change: Optional[int] = None
    ohlc: Optional[Ohlc] = None
    oi: Optional[int] = None
    oi_day_high: Optional[int] = None
    oi_day_low: Optional[int] = None
    sell_quantity: Optional[int] = None
    timestamp: Optional[datetime] = None
    upper_circuit_limit: Optional[float] = None
    volume: Optional[int] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Datum':
        assert isinstance(obj, dict)
        average_price = from_union([from_float, from_none], obj.get("average_price"))
        buy_quantity = from_union([from_int, from_none], obj.get("buy_quantity"))
        depth = from_union([Depth.from_dict, from_none], obj.get("depth"))
        instrument_token = from_union([from_int, from_none], obj.get("instrument_token"))
        last_price = from_union([from_float, from_none], obj.get("last_price"))
        last_quantity = from_union([from_int, from_none], obj.get("last_quantity"))
        last_trade_time = from_union([from_datetime, from_none], obj.get("last_trade_time"))
        lower_circuit_limit = from_union([from_float, from_none], obj.get("lower_circuit_limit"))
        net_change = from_union([from_int, from_none], obj.get("net_change"))
        ohlc = from_union([Ohlc.from_dict, from_none], obj.get("ohlc"))
        oi = from_union([from_int, from_none], obj.get("oi"))
        oi_day_high = from_union([from_int, from_none], obj.get("oi_day_high"))
        oi_day_low = from_union([from_int, from_none], obj.get("oi_day_low"))
        sell_quantity = from_union([from_int, from_none], obj.get("sell_quantity"))
        timestamp = from_union([from_datetime, from_none], obj.get("timestamp"))
        upper_circuit_limit = from_union([from_float, from_none], obj.get("upper_circuit_limit"))
        volume = from_union([from_int, from_none], obj.get("volume"))
        return Datum(average_price, buy_quantity, depth, instrument_token, last_price, last_quantity, last_trade_time, lower_circuit_limit, net_change, ohlc, oi, oi_day_high, oi_day_low, sell_quantity, timestamp, upper_circuit_limit, volume)

    def to_dict(self) -> dict:
        result: dict = {}
        result["average_price"] = from_union([to_float, from_none], self.average_price)
        result["buy_quantity"] = from_union([from_int, from_none], self.buy_quantity)
        result["depth"] = from_union([lambda x: to_class(Depth, x), from_none], self.depth)
        result["instrument_token"] = from_union([from_int, from_none], self.instrument_token)
        result["last_price"] = from_union([to_float, from_none], self.last_price)
        result["last_quantity"] = from_union([from_int, from_none], self.last_quantity)
        result["last_trade_time"] = from_union([lambda x: x.isoformat(), from_none], self.last_trade_time)
        result["lower_circuit_limit"] = from_union([to_float, from_none], self.lower_circuit_limit)
        result["net_change"] = from_union([from_int, from_none], self.net_change)
        result["ohlc"] = from_union([lambda x: to_class(Ohlc, x), from_none], self.ohlc)
        result["oi"] = from_union([from_int, from_none], self.oi)
        result["oi_day_high"] = from_union([from_int, from_none], self.oi_day_high)
        result["oi_day_low"] = from_union([from_int, from_none], self.oi_day_low)
        result["sell_quantity"] = from_union([from_int, from_none], self.sell_quantity)
        result["timestamp"] = from_union([lambda x: x.isoformat(), from_none], self.timestamp)
        result["upper_circuit_limit"] = from_union([to_float, from_none], self.upper_circuit_limit)
        result["volume"] = from_union([from_int, from_none], self.volume)
        return result


@dataclass(slots=True)
class Quote:
    data: Optional[Dict[str, Datum]] = None
    status: Optional[str] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Quote':
        assert isinstance(obj, dict)
        data = from_union([lambda x: from_dict(Datum.from_dict, x), from_none], obj.get("data"))
        status = from_union([from_str, from_none], obj.get("status"))
        return Quote(data, status)

    def to_dict(self) -> dict:
        result: dict = {}
        result["data"] = from_union([lambda x: from_dict(lambda x: to_class(Datum, x), x), from_none], self.data)
        result["status"] = from_union([from_str, from_none], self.status)
        return result


def quote_from_dict(s: Any) -> Quote:
    return Quote.from_dict(s)


def quote_to_dict(x: Quote) -> Any:
    return to_class(Quote, x)

File: JsonSchemaAndStructs/KiteConnect-Objective_C-Structs/Quote/Quote Objective-C (quicktype)/Quote.m
#import "Quote.h"

// Shorthand for simple blocks
#define λ(decl, expr) (^(decl) { return (expr); })

// nil → NSNull conversion for JSON dictionaries
static id NSNullify(id _Nullable x) {
    return (x == nil || x == NSNull.null) ? NSNull.null : x;
}

NS_ASSUME_NONNULL_BEGIN

#pragma mark - Private model interfaces

@interface Quote (JSONConversion)
+ (instancetype)fromJSONDictionary:(NSDictionary *)dict;
- (NSDictionary *)JSONDictionary;
@end

@interface QuoteDatum (JSONConversion)
+ (instancetype)fromJSONDictionary:(NSDictionary *)dict;
- (NSDictionary *)JSONDictionary;
@end

@interface QuoteDepth (JSONConversion)
+ (instancetype)fromJSONDictionary:(NSDictionary *)dict;
- (NSDictionary *)JSONDictionary;
@end

@interface QuoteBuy (JSONConversion)
+ (instancetype)fromJSONDictionary:(NSDictionary *)dict;
- (NSDictionary *)JSONDictionary;
@end

@interface QuoteOhlc (JSONConversion)
+ (instancetype)fromJSONDictionary:(NSDictionary *)dict;
- (NSDictionary *)JSONDictionary;
@end

static id map(id collection, id (^f)(id value)) {
    id result = nil;
    if ([collection isKindOfClass:NSArray.class]) {
        result = [NSMutableArray arrayWithCapacity:[collection count]];
        for (id x in collection) [result addObject:f(x)];
    } else if ([collection isKindOfClass:NSDictionary.class]) {
        result = [NSMutableDictionary dictionaryWithCapacity:[collection count]];
        for (id key in collection) [result setObject:f([collection objectForKey:key]) forKey:key];
    }
    return result;
}

#pragma mark - JSON serialization

Quote *_Nullable QuoteFromData(NSData *data, NSError **error)
{
    @try {
        id json = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:error];
        return *error ? nil : [Quote fromJSONDictionary:json];
    } @catch (NSException *exception) {
        *error = [NSError errorWithDomain:@"JSONSerialization" code:-1 userInfo:@{ @"exception": exception }];
        return nil;
    }
}

Quote *_Nullable QuoteFromJSON(NSString *json, NSStringEncoding encoding, NSError **error)
{
    return QuoteFromData([json dataUsingEncoding:encoding], error);
}

NSData *_Nullable QuoteToData(Quote *, NSError **error)
{
    @try {
        id json = [ JSONDictionary];
        NSData *data = [NSJSONSerialization dataWithJSONObject:json options:kNilOptions error:error];
        return *error ? nil : data;
    } @catch (NSException *exception) {
        *error = [NSError errorWithDomain:@"JSONSerialization" code:-1 userInfo:@{ @"exception": exception }];
        return nil;
    }
}

NSString *_Nullable QuoteToJSON(Quote *, NSStringEncoding encoding, NSError **error)
{
    NSData *data = QuoteToData(, error);
    return data ? [[NSString alloc] initWithData:data encoding:encoding] : nil;
}

@implementation Quote
+ (NSDictionary<NSString *, NSString *> *)properties
{
    static NSDictionary<NSString *, NSString *> *properties;
    return properties = properties ? properties : @{
        @"data": @"data",
        @"status": @"status",
    };
}

+ (_Nullable instancetype)fromData:(NSData *)data error:(NSError *_Nullable *)error
{
    return QuoteFromData(data, error);
}

+ (_Nullable instancetype)fromJSON:(NSString *)json encoding:(NSStringEncoding)encoding error:(NSError *_Nullable *)error
{
    return QuoteFromJSON(json, encoding, error);
}

+ (instancetype)fromJSONDictionary:(NSDictionary *)dict
{
    return dict ? [[Quote alloc] initWithJSONDictionary:dict] : nil;
}

- (instancetype)initWithJSONDictionary:(NSDictionary *)dict
{
    if (self = [super init]) {
        [self setValuesForKeysWithDictionary:dict];
        _data = map(_data, λ(id x, [QuoteDatum fromJSONDictionary:x]));
    }
    return self;
}

- (void)setValue:(nullable id)value forKey:(NSString *)key
{
    id resolved = Quote.properties[key];
    if (resolved) [super setValue:value forKey:resolved];
}

- (void)setNilValueForKey:(NSString *)key
{
    id resolved = Quote.properties[key];
    if (resolved) [super setValue:@(0) forKey:resolved];
}

- (NSDictionary *)JSONDictionary
{
    id dict = [[self dictionaryWithValuesForKeys:Quote.properties.allValues] mutableCopy];

    // Map values that need translation
    [dict addEntriesFromDictionary:@{
        @"data": NSNullify(map(_data, λ(id x, [x JSONDictionary]))),
    }];

    return dict;
}

- (NSData *_Nullable)toData:(NSError *_Nullable *)error
{
    return QuoteToData(self, error);
}

- (NSString *_Nullable)toJSON:(NSStringEncoding)encoding error:(NSError *_Nullable *)error
{
    return QuoteToJSON(self, encoding, error);
}
@end

@implementation QuoteDatum
+ (NSDictionary<NSString *, NSString *> *)properties
{
    static NSDictionary<NSString *, NSString *> *properties;
    return properties = properties ? properties : @{
        @"average_price": @"averagePrice",
        @"buy_quantity": @"buyQuantity",
        @"depth": @"depth",
        @"instrument_token": @"instrumentToken",
        @"last_price": @"lastPrice",
        @"last_quantity": @"lastQuantity",
        @"last_trade_time": @"lastTradeTime",
        @"lower_circuit_limit": @"lowerCircuitLimit",
        @"net_change": @"netChange",
        @"ohlc": @"ohlc",
        @"oi": @"oi",
        @"oi_day_high": @"oiDayHigh",
        @"oi_day_low": @"oiDayLow",
        @"sell_quantity": @"sellQuantity",
        @"timestamp": @"timestamp",
        @"upper_circuit_limit": @"upperCircuitLimit",
        @"volume": @"volume",
    };
}

+ (instancetype)fromJSONDictionary:(NSDictionary *)dict
{
    return dict ? [[QuoteDatum alloc] initWithJSONDictionary:dict] : nil;
}

- (instancetype)initWithJSONDictionary:(NSDictionary *)dict
{
    if (self = [super init]) {
        [self setValuesForKeysWithDictionary:dict];
        _depth = [QuoteDepth fromJSONDictionary:(id)_depth];
        _ohlc = [QuoteOhlc fromJSONDictionary:(id)_ohlc];
    }
    return self;
}

- (void)setValue:(nullable id)value forKey:(NSString *)key
{
    id resolved = QuoteDatum.properties[key];
    if (resolved) [super setValue:value forKey:resolved];
}

- (void)setNilValueForKey:(NSString *)key
{
    id resolved = QuoteDatum.properties[key];
    if (resolved) [super setValue:@(0) forKey:resolved];
}

- (NSDictionary *)JSONDictionary
{
    id dict = [[self dictionaryWithValuesForKeys:QuoteDatum.properties.allValues] mutableCopy];

    // Rewrite property names that differ in JSON
    for (id jsonName in QuoteDatum.properties) {
        id propertyName = QuoteDatum.properties[jsonName];
        if (![jsonName isEqualToString:propertyName]) {
            dict[jsonName] = dict[propertyName];
            [dict removeObjectForKey:propertyName];
        }
    }

    // Map values that need translation
    [dict addEntriesFromDictionary:@{
        @"depth": NSNullify([_depth JSONDictionary]),
        @"ohlc": NSNullify([_ohlc JSONDictionary]),
    }];

    return dict;
}
@end

@implementation QuoteDepth
+ (NSDictionary<NSString *, NSString *> *)properties
{
    static NSDictionary<NSString *, NSString *> *properties;
    return properties = properties ? properties : @{
        @"buy": @"buy",
        @"sell": @"sell",
    };
}

+ (instancetype)fromJSONDictionary:(NSDictionary *)dict
{
    return dict ? [[QuoteDepth alloc] initWithJSONDictionary:dict] : nil;
}

- (instancetype)initWithJSONDictionary:(NSDictionary *)dict
{
    if (self = [super init]) {
        [self setValuesForKeysWithDictionary:dict];
        _buy = map(_buy, λ(id x, [QuoteBuy fromJSONDictionary:x]));
        _sell = map(_sell, λ(id x, [QuoteBuy fromJSONDictionary:x]));
    }
    return self;
}

- (void)setValue:(nullable id)value forKey:(NSString *)key
{
    id resolved = QuoteDepth.properties[key];
    if (resolved) [super setValue:value forKey:resolved];
}

- (void)setNilValueForKey:(NSString *)key
{
    id resolved = QuoteDepth.properties[key];
    if (resolved) [super setValue:@(0) forKey:resolved];
}

- (NSDictionary *)JSONDictionary
{
    id dict = [[self dictionaryWithValuesForKeys:QuoteDepth.properties.allValues] mutableCopy];

    // Map values that need translation
    [dict addEntriesFromDictionary:@{
        @"buy": NSNullify(map(_buy, λ(id x, [x JSONDictionary]))),
        @"sell": NSNullify(map(_sell, λ(id x, [x JSONDictionary]))),
    }];

    return dict;
}
@end

@implementation QuoteBuy
+ (NSDictionary<NSString *, NSString *> *)properties
{
    static NSDictionary<NSString *, NSString *> *properties;
    return properties = properties ? properties : @{
        @"orders": @"orders",
        @"price": @"price",
        @"quantity": @"quantity",
    };
}

+ (instancetype)fromJSONDictionary:(NSDictionary *)dict
{
    return dict ? [[QuoteBuy alloc] initWithJSONDictionary:dict] : nil;
}

- (instancetype)initWithJSONDictionary:(NSDictionary *)dict
{
    if (self = [super init]) {
        [self setValuesForKeysWithDictionary:dict];
    }
    return self;
}

- (void)setValue:(nullable id)value forKey:(NSString *)key
{
    id resolved = QuoteBuy.properties[key];
    if (resolved) [super setValue:value forKey:resolved];
}

- (void)setNilValueForKey:(NSString *)key
{
    id resolved = QuoteBuy.properties[key];
    if (resolved) [super setValue:@(0) forKey:resolved];
}

- (NSDictionary *)JSONDictionary
{
    return [self dictionaryWithValuesForKeys:QuoteBuy.properties.allValues];
}
@end

@implementation QuoteOhlc
+ (NSDictionary<NSString *, NSString *> *)properties
{
    static NSDictionary<NSString *, NSString *> *properties;
    return properties = properties ? properties : @{
        @"close": @"close",
        @"high": @"high",
        @"low": @"low",
        @"open": @"open",
    };
}

+ (instancetype)fromJSONDictionary:(NSDictionary *)dict
{
    return dict ? [[QuoteOhlc alloc] initWithJSONDictionary:dict] : nil;
}

- (instancetype)initWithJSONDictionary:(NSDictionary *)dict
{
    if (self = [super init]) {
        [self setValuesForKeysWithDictionary:dict];
    }
    return self;
}

- (void)setValue:(nullable id)value forKey:(NSString *)key
{
    id resolved = QuoteOhlc.properties[key];
    if (resolved) [super setValue:value forKey:resolved];
}

- (void)setNilValueForKey:(NSString *)key
{
    id resolved = QuoteOhlc.properties[key];
    if (resolved) [super setValue:@(0) forKey:resolved];
}

- (NSDictionary *)JSONDictionary
{
    return [self dictionaryWithValuesForKeys:QuoteOhlc.properties.allValues];
}
@end

NS_ASSUME_NONNULL_END
